<img src=https://www.meme-arsenal.com/memes/5f0ccdd82eda80ded9db721a8746f804.jpg><br>
<hr style="border: 10px solid green;" />

## <span style="color:green">Финальный комментарий<span>
Юрий, спасибо за доработки🤝

Не все оказались верными, посмотри мой комментарий [тестирование модели](#21-bullet) - оставим на самостоятельное исправление

Поздравляю, проект принят👍

<hr style="border: 2px solid green;" />

## Комментарий ревьювера

Юрий, привет! 

🤚Меня зовут, Олег Юрьев, я буду проверять твой проект. Если не против, я буду обращаться к тебе на _ТЫ_ . Если это неприемлемо, обязательно напиши мне в комментариях - мы прейдем на _ВЫ_ .

🤝Спасибо, что сдал проект вовремя. Тобой проделана большая работа, молодец! 

✏️В этой тетрадке ты найдешь мои комментарии.

Комментарии будут выделены:

<div style="border:solid green 2px; padding: 20px"> <b>Отличная работа</b><br>
    Безупречная работа</div>       
<br>
<div style="border:solid orange 2px; padding: 20px"> <b>Что можно улучшить</b><br>
    Ничего дорабатывать не надо, данные пункты работают нормально, но есть способы улучшить. Несколько таких комментариев допустимы, но более - придется некоторые доработать. Большое кол-во мелких замечаний и помарок сказываются на общем впечатлении от работы.</div>   
<br>
<div style="border:solid red 2px; padding: 20px"> <b>Что необходимо доработать</b><br>
    Такие пункты необходимо исправить в первую очередь, чтобы я смог принять проект.</div>
 
Не удаляй комментарии, так будем следить за динамикой 📈
   
Давай будем в диалоге - основные цели в диалоге:
- обмен опытом (да я тоже учусь у студентов👨‍🎓)
- выполнить проект еще лучше
- помочь тебе стать аналитиком

Поэтому чтобы мне проще было проверять повторно, а тебе контролировать процесс ревью, оставляй свои комментарии и выделяй их. 
    
<div class="alert alert-info" style="border:solid blue 2px; padding: 20px"> <b>Комментарий студента:</b> Например, вот так.</div>
   
Поехали! :)

<hr style="border: 2px solid green;" />

# Рекомендация тарифов

В вашем распоряжении данные о поведении клиентов, которые уже перешли на эти тарифы (из проекта курса «Статистический анализ данных»). Нужно построить модель для задачи классификации, которая выберет подходящий тариф. Предобработка данных не понадобится — вы её уже сделали.

Постройте модель с максимально большим значением *accuracy*. Чтобы сдать проект успешно, нужно довести долю правильных ответов по крайней мере до 0.75. Проверьте *accuracy* на тестовой выборке самостоятельно.

## 1. Откройте и изучите файл

In [2]:
import pandas as pd

from sklearn.model_selection import train_test_split

from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression

from sklearn.metrics import accuracy_score

<div style="border:solid green 2px; padding: 20px"> <b>Отличная работа</b><br>
Классно, когда все импорты в одном месте. Если чего не будет, например, коллега увидит это сразу, а не будет ожидать. Некоторые задачи могут решаться довольно долго</div>  

In [3]:
users_behavior = pd.read_csv('/datasets/users_behavior.csv')

In [4]:
users_behavior.shape

(3214, 5)

In [5]:
users_behavior.head()

calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0

В датасете целевой признак is_ultra. Будем строить модель предсказывающую 1 - Ультра, 0 - Смарт. Целевой признак - категориальный.

In [6]:
users_behavior.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
calls       3214 non-null float64
minutes     3214 non-null float64
messages    3214 non-null float64
mb_used     3214 non-null float64
is_ultra    3214 non-null int64
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


<div style="border:solid orange 2px; padding: 20px"> <b>Что можно улучшить</b><br>
Рекомендовал бы не ограничивать обзор данных, даже с учетом того, что он подготовлен на стадии проекта <b>Статистический анализ</b> - помним про принцип <b>Garbage in - garbage out</b>. Легче в самом начале увидеть несоответствие, чем потратив силы и время.<br>
Чтобы тратить меньше времени на обзор, можно автоматизировать этот процесс:<br>
<ul>
    <li> собрать известные методы в функцию (например, head\tail\describe\info)
    <li> воспользоваться какой-либо библиотекой, например, <a href="https://github.com/pandas-profiling/pandas-profiling">pandas_profiling</a>, статья про неё <a href="https://habr.com/ru/company/ruvds/blog/451478/">на хабре</a>
    <li> или провести увлекательный процесс творчества и изучения, чтобы создать свой собственный пакет - когда-то у меня получилось <a href="https://github.com/urevoleg/data-view">data-view</a> и пакет <a href-"https://pypi.org/project/data-view/">на pypi</a></ul></div>   

## 2. Разбейте данные на выборки

In [5]:
# 60% отводим под обучающую выборку
df_train, df_valid = train_test_split(users_behavior, test_size=0.4, random_state=12345)
# половину из оставшихся 40% отдаем на валидационную, другую половину – на тестовую
df_valid, df_test = train_test_split(df_valid, test_size=0.5, random_state=12345)

Мы получили выборки по 60 20 и 20 процентов. Сначала проверим модель на валидационном наборе данных, затем на тестовом.

<div style="border:solid green 2px; padding: 20px"> <b>Отличная работа</b><br>
Выборки разбиты корректно</div>       
<br><div style="border:solid orange 2px; padding: 20px"> <b>Что можно улучшить</b><br>
В комментарии ошибка, так как 60+40+20 = 120%<br></div> 
<div class="alert alert-info" style="border:solid blue 2px; padding: 20px"> <b>Комментарий студента:</b> Исправил.</div>

<div style="border:solid green 2px; padding: 20px"> <b>Отличная работа 2</b><br>👍</div> 

## 3. Исследуйте модели

Для решения задачи классификации применим данные модели и выберем лучшую.

дерево решений / decision tree
случайный лес / random forest
логистическую регрессию / logistic regression

In [6]:
features = users_behavior.drop(['is_ultra'], axis=1)
target = users_behavior['is_ultra']
#целевые признаки обучающей и валидационных выборок
features_train = df_train.drop(['is_ultra'], axis=1)
target_train = df_train['is_ultra']

features_valid = df_valid.drop(['is_ultra'], axis=1)
target_valid = df_valid['is_ultra']

In [7]:
#Функция оценивающая качество модели.
def accuracy_score_valid(model):
    predictions_valid = model.predict(features_valid)
    return accuracy_score(target_valid, predictions_valid)

### Дерево решений

In [8]:
# Влияет ли глубина дерева на оценку правильности?
for max_depth in range(1, 21, 2):
    model = DecisionTreeClassifier(max_depth=max_depth, random_state=12345)
    model.fit(features_train, target_train)
    print(f"max_depth = {max_depth}:\t{accuracy_score_valid(model)}")

max_depth = 1:	0.7542768273716952
max_depth = 3:	0.7853810264385692
max_depth = 5:	0.7791601866251944
max_depth = 7:	0.7822706065318819
max_depth = 9:	0.7822706065318819
max_depth = 11:	0.7620528771384136
max_depth = 13:	0.7558320373250389
max_depth = 15:	0.7465007776049767
max_depth = 17:	0.7356143079315708
max_depth = 19:	0.7278382581648523


Судя по валидационной выборке, дерево решений имеет самую высокую оценку правильности 0.7822706065318819, когда задан гиперпараметр глубины равен 7 и 9.

### Случайный лес

In [9]:
# Гиперпараметр глубины возьмем из пред. задания
# Кол-во деревьев для нашего случайного леса будем искать в диапазоне от 10 до 100 с шагом 10.
for estim in range(10, 101, 10):
    model = RandomForestClassifier(n_estimators=estim, max_depth=7, random_state=12345)
    model.fit(features_train, target_train)
    print(f"n_estimators = {estim}:\t{accuracy_score_valid(model)}")

n_estimators = 10:	0.7947122861586314
n_estimators = 20:	0.8009331259720062
n_estimators = 30:	0.80248833592535
n_estimators = 40:	0.80248833592535
n_estimators = 50:	0.80248833592535
n_estimators = 60:	0.7993779160186625
n_estimators = 70:	0.8009331259720062
n_estimators = 80:	0.7993779160186625
n_estimators = 90:	0.7993779160186625
n_estimators = 100:	0.8009331259720062


Разница есть, но она небольшая

<div style="border:solid orange 2px; padding: 20px"> <b>Что можно улучшить</b><br>
Постараюсь, пояснить маленькую ошибку при определении оптимальных параметров для случайного леса<br>
Представим, что на рисунке ниже изображена кривая обучения модели, по осям Х и Y отложены два гиперпарамера (кол-во деревьев и глубина), по оси Z метрика accuracy (не обращаем внимание на значения, представленные на рисунке). Используя раздельные циклы для 2 гиперпараметров (в твоем случае, ты жестко зафиксировал глубину, выбрав её из дерева решения), ты оказываешься в разных максимумах кривой, при этом совместно они дают результат далекий от ожидаемого. Поэтому перебирать нужно параметры вложенными циклами, чтобы отыскать лучшую пару. Надеюсь, немного прояснил 
    <br>
<img src=http://storage.yandexcloud.net/public-bucket-6/%D0%BA%D1%80%D0%B8%D0%B2%D0%B0%D1%8F.png><br><br>
Узнать больше <a href="https://dyakonov.org/tag/random-forest/">Про параметры RandomForest</a></div>

### Логистическая регрессия

In [10]:
def logistic_regression(solver='liblinear'):
    model = LogisticRegression(solver=solver, random_state=12345)
    model.fit(features_train, target_train)
    return accuracy_score_valid(model)

In [11]:
logistic_regression()

0.7589424572317263

Логистическая регрессия показывает худшие результаты

<div style="border:solid orange 2px; padding: 20px"> <b>Что можно улучшить</b><br>
По секрету, у логистической регрессии тоже есть гиперпараметры, например, <code>C</code><br><br>
Перебирая гиперпараметры не всегда будешь выводить результат через <code>print</code> или ждать неизвестного - хочется следить за прогрессом - в этом может помочь библиотека <a href="https://github.com/tqdm/tqdm">tqdm</a></div>  

In [12]:
from tqdm.notebook import tqdm
import time
for depth in tqdm(range(10), desc='depth'):
    for estim in tqdm(range(10, 101, 10), desc='estim', leave=False):
        time.sleep(0.05)

<div class="alert alert-info" style="border:solid blue 2px; padding: 20px"> <b>Комментарий студента:</b> Спасибо за ценную инфу!</div>

<div style="border:solid green 2px; padding: 20px"> <b>Отличная работа 2</b><br>👍</div> 

## 4. Проверьте модель на тестовой выборке

В результате экспериментов, мы поняли, что лучший результат показывает случайный лес решений. Параметр глубины 7 и кол-во деревьев 70 самый оптимальный вариант. Проверим его на тестовой выборке.

In [13]:
features_test = df_test.drop(['is_ultra'], axis=1)
target_test = df_test['is_ultra']

model = RandomForestClassifier(n_estimators=70, max_depth=7, random_state=12345)
model.fit(features_train, target_train)

predictions_test = model.predict(features_test)
accuracy_score(target_test, predictions_test)



0.80248833592535

In [14]:
test = model.predict(features)
accuracy_score(target,test)

0.8391412570006223

Значения практически равны! Это говорит о качестве модели. Проблемы с переобучением нет, поэтому можно сделать вывод, что мы правильно выполнили поставленную задачу.

<a class="anchor" id="21-bullet">

<div style="border:solid red 2px; padding: 20px"> <b>Что необходимо доработать</b><br>
Классно, что получилась отличная и устойчивая модель. Чтобы еще более корректно проверить модель, необходимо объединить выборки <code>train\valid</code> и обучить модель и проверить метрику на test</div>
<div class="alert alert-info" style="border:solid blue 2px; padding: 20px"> <b>Комментарий студента:</b> Сделал features и target по выборке из общего датасета</div>
<br>
<div style="border:solid green 2px; padding: 20px"> <b>Комментарий 2</b><br>
Смотри, ты только протестировал модель на всех данных (из них модель видела 60% и не видела 40%). Нужно обучить модель на 80% данных (train\valid) и протестировать на 20% (test).<br>
    В этом поможет <code>pd.concat([df_train, df_valid])</code></div>   

## 5. (бонус) Проверьте модели на адекватность

Посчитаем количество перешедших на тариф "смарт" к общему количеству объектов в выборке. Посчитаю, сколько людей перешло на тариф смарт, по отношению к общему кол-ву объектов.

In [15]:
(users_behavior['is_ultra']==0).sum() / users_behavior.shape[0]

0.693528313627878

Доля равна ~0.69, в то время как полученная нами модель имеет accuracy ~0.79. Мы можем сделать вывод, наша модель считается адекватной для работы.

<div style="border:solid green 2px; padding: 20px"> <b>Отличная работа</b><br>
Молодец, действительно мы предсказываем много лучше константы</div>   

<hr style="border: 2px solid green;" />

## Общий комментарий
Спасибо за проделанную работу🤝

Работа выполнена хорошо. Ты правильно разделил данные на выборки, обучая модель, нашел лучшие гиперпараметры и убедился в адекватности модели.

Остались моменты, которые нужно поправить:
- корректно протестировать модель
- (рекомендую) поискать лучшие гиперпараметры для случайного леса, перебирая и глубину и кол-во деревьев
- очень хочется увидеть общий вывод, его можно построить следующим образом:
    - краткий обзор выполненных работ (выполнили, сделали, рассчитали)
    - основные выводы (ответы на вопросы поставленные перед проектом)
    - рекомендации (если они (рекомендации) возможны)
    
Исправляй и отправляй на проверку🛠

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x] Jupyter Notebook открыт
- [x] Весь код исполняется без ошибок
- [x] Ячейки с кодом расположены в порядке исполнения
- [x] Выполнено задание 1: данные загружены и изучены
- [x] Выполнено задание 2: данные разбиты на три выборки
- [x] Выполнено задание 3: проведено исследование моделей
    - [x] Рассмотрено больше одной модели
    - [x] Рассмотрено хотя бы 3 значения гипепараметров для какой-нибудь модели
    - [x] Написаны выводы по результатам исследования
- [x] Выполнено задание 3: Проведено тестирование
- [x] Удалось достичь accuracy не меньше 0.75
